drivers/usb/usbuhci_new/usbuhci.c

```diff
@@ -0,0 +1,2897 @@
+{
+    DPRINT("TD              - %p\n", TD);
+    DPRINT("NextElement     - %p\n", TD->HwTD.NextElement);
+    DPRINT("ControlStatus   - %08X\n", TD->HwTD.ControlStatus.AsULONG);
+    DPRINT("Token           - %p\n", TD->HwTD.Token.AsULONG);
+if (TD->HwTD.Buffer)
+    DPRINT("Buffer          - %p\n", TD->HwTD.Buffer);
+
+if (TD->SetupPacket.bmRequestType.B)
+    DPRINT("bmRequestType   - %02X\n", TD->SetupPacket.bmRequestType.B);
+if (TD->SetupPacket.bRequest)
+    DPRINT("bRequest        - %02X\n", TD->SetupPacket.bRequest);
+if (TD->SetupPacket.wValue.W)
+    DPRINT("wValue          - %04X\n", TD->SetupPacket.wValue.W);
+if (TD->SetupPacket.wIndex.W)
+    DPRINT("wIndex          - %04X\n", TD->SetupPacket.wIndex.W);
+if (TD->SetupPacket.wLength)
+    DPRINT("wLength         - %04X\n", TD->SetupPacket.wLength);
+
+    DPRINT("PhysicalAddress - %p\n", TD->PhysicalAddress);
+    DPRINT("Flags           - %X\n", TD->Flags);
+    DPRINT("NextHcdTD       - %p\n", TD->NextHcdTD);
+    DPRINT("UhciTransfer    - %p\n", TD->UhciTransfer);
+}
+
+VOID
+NTAPI
+UhciFixDataToggle(IN PUHCI_EXTENSION UhciExtension,
+                  IN PUHCI_ENDPOINT UhciEndpoint,
+                  IN PUHCI_HCD_TD TD,
+                  IN BOOL DataToggle)
+{
+    DPRINT_UHCI("UhciFixDataToggle: UhciExtension - %p, UhciEndpoint - %p, DataToggle - %X\n",
+                UhciExtension,
+                UhciEndpoint,
+                DataToggle);
+
+    while (TD)
+    {
+        TD->HwTD.Token.DataToggle = (TD->HwTD.Token.DataToggle == FALSE);
+        DataToggle = (DataToggle == FALSE);
+
+        TD = TD->NextHcdTD;
+    }
+
+    UhciEndpoint->DataToggle = DataToggle;
+}
+
+VOID
+NTAPI
+UhciCleanupFrameListEntry(IN PUHCI_EXTENSION UhciExtension,
+                          IN ULONG Index)
+{
+    PUHCI_HC_RESOURCES UhciResources;
+    ULONG_PTR PhysicalAddress;
+    ULONG HeadIdx;
+
+    UhciResources = UhciExtension->HcResourcesVA;
+
+    if (Index == 0)
+    {
+        PhysicalAddress = UhciExtension->StaticTD->PhysicalAddress;
+
+        UhciResources->FrameList[0] = PhysicalAddress |
+                                      UHCI_FRAME_LIST_POINTER_TD;
+    }
+    else
+    {
+        HeadIdx = (INTERRUPT_ENDPOINTs - ENDPOINT_INTERRUPT_32ms) +
+                  (Index & (ENDPOINT_INTERRUPT_32ms - 1));
+
+        PhysicalAddress = UhciExtension->IntQH[HeadIdx]->PhysicalAddress;
+
+        UhciResources->FrameList[Index] = PhysicalAddress |
+                                          UHCI_FRAME_LIST_POINTER_QH;
+    }
+}
+
+VOID
+NTAPI 
+UhciCleanupFrameList(IN PUHCI_EXTENSION UhciExtension,
+                     IN BOOLEAN IsAllEntries)
+{
+    ULONG NewFrameNumber;
+    ULONG OldFrameNumber;
+    ULONG ix;
+
+    DPRINT_UHCI("UhciCleanupFrameList: [%p] All - %x\n",
+                UhciExtension, IsAllEntries);
+
+    if (InterlockedIncrement(&UhciExtension->LockFrameList) != 1)
+    {
+        InterlockedDecrement(&UhciExtension->LockFrameList);
```

Comment:
This lock seems odd. It protects code rather than data, which is a red flag.
If a second thread can get here and bail out while the first thread is already done walking the list but hasn't released the lock yet, the second cleanup would do nothing, even though there may be work to be done. Is this acceptable for this function? That might deserve a comment if so.

Reply:
This is clearing the list for isochronous transfers. This is also not implemented yet, so it may look strange. I can make them as a comment, so as not to forget about them. Or?

Reply:
A comment sounds fine to me.